experimental/devices/inky/inky.go

```diff
@@ -0,0 +1,327 @@
+	// Red/Transparent pixels.
+	red := make([]bool, rows*cols)
+	for x := b.Min.X; x < b.Max.X; x++ {
+		for y := b.Min.Y; y < b.Max.Y; y++ {
+			i := x*cols + y
+			srcX := x
+			srcY := b.Max.Y - y - 1
+			r, g, b, _ := d.ColorModel().Convert(src.At(srcX, srcY)).RGBA()
+			if r >= 0x8000 && g >= 0x8000 && b >= 0x8000 {
+				white[i] = true
+			} else if r >= 0x8000 {
+				// Red pixels also need white behind them.
+				white[i] = true
+				red[i] = true
+			}
+		}
+	}
+
+	bufA, _ := pack(white)
+	bufB, _ := pack(red)
+	return d.update(borderColor[d.border], bufA, bufB)
+}
+
+func (d *Dev) update(border byte, black []byte, red []byte) error {
+	if err := d.reset(); err != nil {
+		return err
+	}
+
+	if err := d.sendCommand(0x74, []byte{0x54}); err != nil { // Set Analog Block Control.
+		return err
+	}
+	if err := d.sendCommand(0x7e, []byte{0x3b}); err != nil { // Set Digital Block Control.
+		return err
+	}
+
+	r := make([]byte, 3)
+	binary.LittleEndian.PutUint16(r, rows)
+	if err := d.sendCommand(0x01, r); err != nil { // Gate setting
+		return err
+	}
+
+	init := []struct {
+		cmd  byte
+		data []byte
+	}{
+		{0x03, []byte{0x10, 0x01}}, // Gate Driving Voltage.
+		{0x3a, []byte{0x07}},       // Dummy line period
+		{0x3b, []byte{0x04}},       // Gate line width
+		{0x11, []byte{0x03}},       // Data entry mode setting 0x03 = X/Y increment
+		{0x04, nil},                // Power on
+		{0x2c, []byte{0x3c}},       // VCOM Register, 0x3c = -1.5v?
+		{0x3c, []byte{0x00}},
+		{0x3c, []byte{byte(border)}}, // Border colour
+	}
+
+	for _, c := range init {
+		if err := d.sendCommand(c.cmd, c.data); err != nil {
+			return err
+		}
+	}
+
+	switch d.color {
+	case Black:
+		if err := d.sendCommand(0x32, blackLUT[:]); err != nil {
+			return err
+		}
+	case Red:
+		if err := d.sendCommand(0x32, redLUT[:]); err != nil {
+			return err
+		}
+	case Yellow:
+		if err := d.sendCommand(0x04, []byte{0x07}); err != nil { // Set voltage of VSH and VSL.
+			return err
+		}
+		if err := d.sendCommand(0x32, yellowLUT[:]); err != nil {
+			return err
+		}
+	}
+
+	h := make([]byte, 4)
+	binary.LittleEndian.PutUint16(h[2:], rows)
+	write := []struct {
+		cmd  byte
+		data []byte
+	}{
+		{0x44, []byte{0x00, cols/8 - 1}}, // Set RAM X Start/End
+		{0x45, h},                        // Set RAM Y Start/End
+		{0x43, []byte{0x00}},
+
+		{0x4e, []byte{0x00}},
+		{0x4f, []byte{0x00, 0x00}},
+		{0x24, black},
+
+		{0x43, []byte{0x00}},
+		{0x4f, []byte{0x00, 0x00}},
+		{0x26, red},
+
+		{0x22, []byte{0xc7}},
+	}
+
+	for _, c := range write {
+		if err := d.sendCommand(c.cmd, c.data); err != nil {
+			return err
+		}
+	}
+
+	d.busy.In(gpio.PullUp, gpio.FallingEdge)
+	defer d.busy.In(gpio.PullUp, gpio.NoEdge)
+	if err := d.sendCommand(0x20, nil); err != nil {
+		return err
+	}
+
+	d.busy.WaitForEdge(-1)
+
+	if err := d.sendCommand(0x10, []byte{0x01}); err != nil { // Enter deep sleep.
+		return err
+	}
+	return nil
+}
+
+func (d *Dev) reset() error {
+	d.r.Out(gpio.Low)
```

Comment:
Please consistently check for errors.

Reply:
Is it worth checking errors on every gpio write?

Reply:
That's a fair question. In theory I'd say no. In practice, I do because of these reasons:

- periph's code is meant to be as the reference for high quality code.
- People tend to copy paste code, and as such, the code should do the right thing.
- You can't know what failure mode will happen.

Here it's a special case where we know the possibility of the hardware setup is fairly constrained, but people are more creative than that.

Imagine someone hacks up a setup with a FT232H (connected via USB) on their Windows workstation and connections up the wires to drive an inky from there. It's kinda cool when you think about it. There's 6 data wires to connect plus power. In this setup, if the user unplug the FT232H by accident, the functions will start failing at runtime.

Reply:
That's fair. What do you think about the case where you want to do something like this?
```go
pin.In(gpio.PullUp, gpio.FallingEdge)
defer pin.In(gpio.PullUp, gpio.NoEdge)
```
I guess it's a bit like doing `defer file.Close()` on a writable file; i.e., you probably shouldn't use `defer` and maybe use a named return for the `error`?